### ISSUE TITLE
Linting

### ISSUE THREAD

**Issue by bogdanteleaga** [ipython-contrib/jupyter_contrib_nbextensions#913]
Just got started with extensions and it looks to me like a big missing one is a general linting mechanism.

Something like `flake8`/`pylint`/`mypy` for python would be very useful.

**Comment by jcb91** [comment]
Autocorrection is already implemented. You can use [autopep8](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/blob/28e5ea4/src/jupyter_contrib_nbextensions/nbextensions/code_prettify/READMEautopep8.md) for autopep8, or [code_prettify](https://github.com/ipython-contrib/jupyter_contrib_nbextensions/blob/28e5ea4/src/jupyter_contrib_nbextensions/nbextensions/code_prettify/README_code_prettify.md), which by default uses yapf, but is configurable to use any kernel-side linting library.

If what you're after is some way to _mark_ linting errors _without_ automatically fixing them, that's another idea, but one I can see the utility of, and that I don't would be hugely difficult to implement. What exactly is the desired use process that you envisage?

**Comment by bogdanteleaga** [comment]
Yes, I was talking about only signalling errors without fixing them, I am already using code_prettify for yapf. 

Some examples of errors caught by `flake8` are found [here](http://flake8.pycqa.org/en/latest/user/error-codes.html). For `mypy` this is maybe easier to envision. The main idea is to catch not only style/formatting errors which can already be automatically fixed by the tools you mentioned, but also logical errors/type errors which need to be addressed by the user.

An example from atom can be seen below:
![](https://cloud.githubusercontent.com/assets/4278113/8768482/52f975c6-2e3f-11e5-87e4-27c8359fd36c.gif)

It could also be toggled by a keypress instead of happening automatically. Another option would be to display a list of errors at the bottom of the cell. One issue would be that some errors are found only if the linter is ran across all of the code, but I imagine that is not a big roadblock.

**Comment by jcb91** [comment]
Ok, thanks for the info, that definitely sounds possible.

> One issue would be that some errors are found only if the linter is ran across all of the code, but I imagine that is not a big roadblock

Yeah, the one that springs to mind is unused/undefined variables as a result of the definition being in a different cell to usage. The only solutions that come to mind are either disabling those lint warnings, or linting all the code cells in the notebook as a single text, then somehow splitting out the relevant linting errors afterwards, but I'd have to look at a specific linter to think about that, I guess. Let me know if you have another plan! I'll not get around to this till I get back to a computer in a week or so, but it sounds interesting, so worth a look. Also feel free to PR anything you come up with :)

**Comment by bogdanteleaga** [comment]
Mypy probably won't really work properly unless you can do it for the whole
file so the 2nd option sounds better. As far as I know the linter outputs
something like `line-no: error`, so as long as you can map back into the
cell line from the line in the file it should be relatively easy. Assuming
the output format is general enough this might be doable in the
code-prettify style where you can plug in any formatter but I'm not
familiar with ones in different languages.

Unfortunately I can't work with js that efficiently so I don't think I can
help with this one.

Something to keep in mind is that you usually want to be able to configure
it to ignore some errors so that would be a helpful thing to have in the
configurator.

Whenever you can get to it is fine, I'm definitely not in a rush, but I do
believe it will be a nice addition.

**Comment by bogdanteleaga** [comment]
This blogpost tells a story about implementing something similar to this, but using cell magic. Could be relevant.
[http://journalpanic.com/post/spice-up-thy-jupyter-notebooks-with-mypy/](http://journalpanic.com/post/spice-up-thy-jupyter-notebooks-with-mypy/)
